### DISCUSSION TITLE
How do I automerge a PR and add "[ci skip]" to the --merge commit-- message?

### DISCUSSION THREAD

**Discussion by klippx** [renovatebot/renovate#24565] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

actions/checkout@v3

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Given this config

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "platform": "github",
  "extends": ["config:js-app", ":disableDependencyDashboard"],
  "platformAutomerge": true,
  "packageRules": [
    {
      "extends": [
        "monorepo:jest"
      ],
      "matchUpdateTypes": ["patch", "minor"],
      "matchCurrentVersion": "!/^0/",
      "automerge": true
    }
  ],
  "lockFileMaintenance": {
    "enabled": true,
    "automerge": true
  }
}
```

Lets say a PR comes up `chore(deps): update jest monorepo to v29.7.0`. I want 

1. CI to run ✅ . This works, since commit message is `chore(deps): update jest monorepo to v29.7.0`
2. Automerge to apply ✅ 
3. On main: CI to NOT run. ⛔️ 

The problem is CI runs on main and that triggers deployment of the service which is totally undesired for devDependencies updates.
 

### What you tried so far.

I have tried using https://docs.renovatebot.com/configuration-options/#commitbody but this applies it to the PR commit and as such, [1] stops running CI which is incorrect. I need the checks in [1] to run or it wont even know if checks pass so it can automerge.

For this to happen I need the **merge commit** to be `chore(deps): update jest monorepo to v29.7.0 [ci skip]`

Can this be done using platformAutomerge? Or does GH just merge the PR "as is", without possibility for renovate to provide a hook to alter the commit message?

If so, does that limit me to use `platformAutomerge: false` and use some renovate config to merge PR during the scheduled runs and modify the commit message to append "[ci skip]" only when merging? And if so, can you help me find the docs for it?

If not, I guess I have an unsupported use case or do you have any other suggestion how people normally deal with this?

### Relevant debug logs

<details><summary>Logs</summary>

```
Pending
```

</details>

**Comment by rarkins** [comment]
GitHub is in charge of how PRs are converted to commits when platform automerging. I believe it's somewhat configurable but not sure to what extent.

Are you wanting to skip CI when the PR is created, or after it's merged to the base branch, or skip in both cases?

**Reply by klippx** [reply]
When merged to base branch: **SKIP**. The reason is because that will stop deployment of service on dev dependency change, this is what [ci skip] does (more like [cd skip] really).

When PR is created: **DO NOT SKIP**. I still need the test suite to run on branch to see result of required checks and test suite. If anything goes wrong there I do not want to merge this to main.

**Reply by rarkins** [reply]
Next question: how does your CI decide to skip? Is it based on commit message, not PR title or body, right? And is it based only on the first line of the commit message, or on any line of the commit message/body?

**Reply by klippx** [reply]
Yes, only commit message, not PR title/body. 

It is built in by GH, documented here: https://docs.github.com/en/actions/managing-workflow-runs/skipping-workflow-runs

The docs only mention "commit message" but I think it works with commit body as well.

**Reply by rarkins** [reply]
If you want CI to be skipped when a Renovate PR is merged to main, it means you need `[ci skip]` to be in _that_ commit message.

If you take this recent automerged commit in Renovate's own repo by example: https://github.com/renovatebot/renovate/commit/9e6ed574c29842df8143cbe94022b0e59a58fadb

The commit message is `chore(deps): update dependency @types/semver to v7.5.2 (#24557)`. It's been crafted by GitHub to include the PR number in the title. I'm not sure if it's using the PR title or the commitMessage, as they're the same.

Related github community discussion: https://github.com/orgs/community/discussions/15925

**Reply by klippx** [reply]
>  it means you need [ci skip] to be in that commit message.

Precisely, this was my original inquiry: if and how can we do this with renovate? 

When we do this manually we edit the commit message and add `[ci skip]` because we can see the branch was green. We use squash commits. so we can edit the squash commit message via the GH UI.

**Reply by rarkins** [reply]
If you use platformAutomerge, then the answer needs to come from GitHub's capabilities there (and it seems it can't).

If you disable platformAutomerge then Renovate will call GitHub's PR merge API. There the functionality is still largely determined by GitHub, but their API may allow more customization. You would need to check their API docs first to see if/how it's customizable before we can answer if it's possible by Renovate

**Reply by klippx** [reply, marked as answer]
We decided to use `"commitMessageSuffix": "[skip cd]",` and implement our own rules in GH action to not trigger deploy pipelines if this message is present. We also needed to add the bot user to be allowed to bypass branch protection rules (1 approval needed) and then it works fine.
